src/pass/zero_elimination.cc

```diff
@@ -0,0 +1,1719 @@
+                        {combiner->rhs[value_index], zero}});
+  in = SuperSimplify(in);
+
+  return is_const_value(in, 0);
+}
+
+Expr InlineThisCall(const Expr& expr) {
+  if (const Call* op = expr.as<Call>()) {
+    if (op->call_type == Call::CallType::Halide) {
+      if (const ComputeOpNode* op_comp = op->func.as<ComputeOpNode>()) {
+        Array<Var> tensor_axes;
+        for (const auto& var : op_comp->axis) {
+          tensor_axes.push_back(var->var);
+        }
+
+        Stmt inlined = Inline(Evaluate::make(expr), op->func, tensor_axes,
+                              op_comp->body[op->value_index]);
+        if (const ir::Evaluate* ev = inlined.as<ir::Evaluate>()) {
+          // If it is a reduction, clone it
+          return CloneReduction(ev->value);
+        }
+      }
+    }
+  }
+
+  return expr;
+}
+
+Tensor InlineTailCall(const Tensor& tensor) {
+  return op::TransformBody(tensor, InlineThisCall);
+}
+
+class InlineTensorsMutator : public IRMutator {
+ public:
+  explicit InlineTensorsMutator(const Array<Tensor>& inlineable, bool inline_reductions = false)
+      : inline_reductions_(inline_reductions) {
+    for (const Tensor& tensor : inlineable) {
+      inlineable_.emplace(tensor->op.operator->(), tensor->value_index);
+    }
+  }
+
+  Expr Mutate_(const Call* op, const Expr& e) {
+    if (op->call_type == Call::CallType::Halide) {
+      const ComputeOpNode* op_comp = op->func.as<ComputeOpNode>();
+      if (inlineable_.empty() || inlineable_.count({op_comp, op->value_index})) {
+        if (op_comp && (inline_reductions_ || !op_comp->body[0].as<Reduce>())) {
+          Array<Var> tensor_axes;
+          for (const auto& var : op_comp->axis) {
+            tensor_axes.push_back(var->var);
+          }
+
+          Stmt inlined = Inline(Evaluate::make(e), op->func, tensor_axes,
+                                op_comp->body[op->value_index]);
+          if (const ir::Evaluate* ev = inlined.as<ir::Evaluate>()) {
+            // If it is a reduction, clone it
+            return Mutate(ev->value);
+          }
+        }
+      }
+    }
+
+    return e;
+  }
+
+ private:
+  std::set<std::pair<const OperationNode*, int>> inlineable_;
+  bool inline_reductions_;
+};
+
+Expr InlineTensors(const Expr& expr, const Array<Tensor>& inlineable,
+                   bool inline_reductions) {
+  return InlineTensorsMutator(inlineable, inline_reductions).Mutate(expr);
+}
+
+Tensor InlineTensors(const Tensor& tensor, const Array<Tensor>& inlineable,
+                     bool inline_reductions) {
+  auto transformation =
+    [inlineable, inline_reductions](const Expr& e) {
+      return InlineTensorsMutator(inlineable, inline_reductions).Mutate(e); };
+  return op::TransformBody(tensor, transformation);
+}
+
+
+struct NonzeronessConditionResult {
+  Expr cond;
+  Expr value;
+
+  Expr to_expr() const {
+    return SelectElseZero(cond, value);
+  }
+};
+
+class NonzeronessConditionFunctor
```

Comment:
Comments please. I guess my review comments are getting repetitive. :smile:

Reply:
Well, I found a problem in this class concerned with treatment of `if_then_else` (it should be preserved when lifting conditions since the lifted conditions are lifted as selects), so I guess lack of comments correlates with bugs.